diff --git a/js/environment_indicator.js b/js/environment_indicator.js
index 7d73b98e3..b503b0144 100644
--- a/js/environment_indicator.js
+++ b/js/environment_indicator.js
@@ -4,7 +4,7 @@

   Drupal.behaviors.environmentIndicatorSwitcher = {
     attach: function (context, settings) {
-      $('#environment-indicator', context).bind('click', function () {
+      $('#environment-indicator', context).off().on('click', function () {
         $('#environment-indicator .environment-switcher-container', context).slideToggle('fast');
       });
     }
@@ -13,8 +13,8 @@
   Drupal.behaviors.environmentIndicatorToolbar = {
     attach: function (context, settings) {
       if (typeof(settings.environmentIndicator) != 'undefined') {
-        $('#toolbar-bar', context).css('background-color', settings.environmentIndicator.bgColor);
-        $('#toolbar-bar .toolbar-item, #toolbar-bar .toolbar-item a', context).css('color', settings.environmentIndicator.fgColor);
+        // $('#toolbar-bar', context).css('background-color', settings.environmentIndicator.bgColor);
+        // $('#toolbar-bar .toolbar-item, #toolbar-bar .toolbar-item a', context).css('color', settings.environmentIndicator.fgColor);

         // Set environment color for gin_toolbar vertical toolbar.
         if ($('body').hasClass('gin--vertical-toolbar')) {
